add/default-route-table

cdobbyn opened the pull request:
## what
* Manages the default route table created alongside `aws_vpc` resource automatically.

## why
* If not managed there are no identifying features about this resource which is confusing.
* In our case we establish ownership via tags passed into this module, if no tags are present it is difficult to report on ownership

## references
* Closes #109 
* [aws_default_route_table](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/default_route_table)

cdobbyn commented:
/test all

nitrocode commented:
/test all

nitrocode commented on main.tf:
```diff
@@ -100,3 +100,11 @@ resource "aws_vpc_ipv6_cidr_block_association" "default" {
+resource "aws_default_route_table" "default" {
+  count                  = local.enabled ? 1 : 0
+  default_route_table_id = aws_vpc.default.id
+  tags = merge(module.label.tags, {
```
cc: @aknysh @Nuru what do you folks think about this current tag/name structure? Would it be better to do this? Thoughts?

```suggestion
  tags                   = module.label.tags
```

Or have an additional input

```suggestion
  tags                   = merge(module.label.tags, var.tags_default_route)
```

Where tags_default_tags contains default=true.

cdobbyn commented:
@nitrocode I just pushed up a fix for that error reported by terratest/bats. Should succeed now if you try again. I wasn't able to get the test runner to work locally (I'm probably doing something incorrectly), was using various `make` commands to try to get it to work.

When releasing I would recommend placing something release notes indicating the new resource as people who use this may be taken aback by it asking to "add" a resource (that already exists).

Feel free to ping me on slack if you think a tag change is required. Easy enough to implement.

nitrocode commented:
/test all

cdobbyn commented:
@nitrocode once more. I was incorrectly pointing at `id` instead of `default_route_table_id`. Sorry for the extra runs.

I also confirmed that this output value is available through `hashicorp/aws` provider 3.0.0 and onward.

cdobbyn commented:
Just looking at this do you think it's worthwhile also bringing in these?
- [aws_default_security_group](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/default_security_group)
- [aws_default_network_acl](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/default_network_acl)

cdobbyn commented:
I ended up adding 3 booleans for enabling / disabling the management of these resources. This should retain the existing behaviour for people whom don't want this.

I'll add an extra test for this case.

nitrocode commented on examples/complete/main.tf:
```diff
@@ -20,6 +20,17 @@ module "vpc" {
+  context = module.this.context
+}
+
+module "vpc_defaults" {
```
Instead of creating another instance of it in the same root module, we may want to expose the `adopt_` inputs in example/complete, set them to `false` in the current test fixtures tfvars, and create a new set of test fixtures tfvars to use the `_adopt.* = true` and a corresponding terratest.

nitrocode commented on main.tf:
```diff
@@ -100,3 +100,27 @@ resource "aws_vpc_ipv6_cidr_block_association" "default" {
+resource "aws_default_route_table" "default" {
+  count                  = local.enabled && var.adopt_default_route_table ? 1 : 0
+  default_route_table_id = aws_vpc.default[0].default_route_table_id
+  tags = merge(module.label.tags, {
+    Name = join(module.label.delimiter, [module.label.id, "default"])
+  })
```
Please put this in a local such as `default_route_table_tags` and use it here

```suggestion
  tags = local.default_route_table_tags
```

nitrocode commented on variables.tf:
```diff
@@ -160,3 +160,18 @@ variable "ipv6_egress_only_internet_gateway_enabled" {
 }
+variable "adopt_default_route_table" {
+  type        = bool
+  description = "A boolean flag to enable/disable adoption of the default route table"
```
Change `A boolean flag` to `Whether` for each new input please

```suggestion
  description = "Whether to enable/disable adoption of the default route table"
```

nitrocode commented on variables.tf:
```diff
@@ -160,3 +160,18 @@ variable "ipv6_egress_only_internet_gateway_enabled" {
+  description = "A boolean flag to enable/disable adoption of the default route table"
+  default     = false
+}
+variable "adopt_default_network_acl" {
```
```suggestion

variable "adopt_default_network_acl" {
```

nitrocode commented on variables.tf:
```diff
@@ -160,3 +160,18 @@ variable "ipv6_egress_only_internet_gateway_enabled" {
+  description = "A boolean flag to enable/disable adoption of the default network acl"
+  default     = false
+}
+variable "adopt_default_security_group" {
```
```suggestion

variable "adopt_default_security_group" {
```

nitrocode commented on examples/complete/fixtures.defaults.tfvars:
```diff
@@ -0,0 +1,11 @@
+region = "us-east-2"
```
There's already a `fixtures.us-east-2.tfvars`.

Let's call this file `fixtures-default-route-table.defaults.tfvars`

nitrocode commented on examples/complete/fixtures.defaults.tfvars:
```diff
@@ -0,0 +1,11 @@
+stage     = "test"
+name      = "vpc-adopts"
+
+default_security_group_deny_all = true
+default_route_table_no_routes   = true
+default_network_acl_deny_all    = true
```
These inputs will need to be added to `fixtures.us-east-2.tfvars` with `false` values

nitrocode commented on main.tf:
```diff
@@ -11,6 +11,9 @@ locals {
     length(local.additional_cidr_blocks_map) > 0 ? local.additional_cidr_blocks_map : var.ipv4_additional_cidr_block_associations
   ) : {}
+  default_adoption_tags = merge(module.label.tags, {
+    Name = join(module.label.delimiter, [module.label.id, "default"])
```
@aknysh  The more I look at this, the more this seems like an `attribute`. I wonder if we should instantiate the null label module again.

nitrocode commented on main.tf:
```diff
@@ -11,6 +11,9 @@ locals {
+  })
 }
 
 module "label" {
```
It's odd that this label does absolutely nothing but has been created.

If we added `attributes = ["default"]` here, we can set the below tags to be `module.label.tags` and remove the local above.

nitrocode replied:
Hmm, but it could cause other issues (such as recreations of resources i.e. no backward compatibility) depending on where this label is used...

nitrocode commented on test/src/examples_complete_test.go:
```diff
@@ -93,3 +93,33 @@ func TestExamplesCompleteDisabled(t *testing.T) {
+
+	rootFolder := "../../"
+	terraformFolderRelativeToRoot := "examples/complete"
+	varFiles := []string{"fixtures.defaults.tfvars"}
```
```suggestion
	varFiles := []string{"fixtures-default-route-table.defaults.tfvars"}
```

cdobbyn replied:
This is testing configuration of the default route table, security group, and network acl.

nitrocode commented on test/src/examples_complete_test.go:
```diff
@@ -93,3 +93,33 @@ func TestExamplesCompleteDisabled(t *testing.T) {
+		VarFiles: varFiles,
+		Vars: map[string]interface{}{
+			"attributes": attributes,
+			"enabled":    "false",
```
Looks like you copied from the `TestExamplesCompleteDisabled` test instead of `TestExamplesComplete`

cdobbyn replied:
Oh you're right. 1m

nitrocode replied:
@cdobbyn maybe we should set those inputs as `true` and put it in the existing fixtures so a new test doesnt have to be added ?

cdobbyn replied:
@nitrocode I was thinking of coverage over both cases, but if you're comfortable with that I can do that.

cdobbyn commented:
I updated some of the naming as per your suggestion @nitrocode. I didn't do it exactly as you named it as this test actually covers 3 things:
- default security group (already existed in this module)
- default network acl (new)
- default route table (new)

nitrocode commented on test/src/examples_complete_test.go:
```diff
@@ -93,3 +93,32 @@ func TestExamplesCompleteDisabled(t *testing.T) {
+		},
+	}
+
+	// At the end of the test, run `terraform destroy` to clean up any resources that were created
```
If we're going to create a new test, we should validate something in it like we do in the other one.

nitrocode commented:
@cdobbyn did you see this comment https://github.com/cloudposse/terraform-aws-vpc/pull/110#discussion_r971267496

cdobbyn commented:
@nitrocode I updated as per https://github.com/cloudposse/terraform-aws-vpc/pull/110#discussion_r971267496.

I removed the standalone test and made the complete one set all values to true.

The "Summary" action on here seems to be a subscription issue and not a code one (it keeps throwing me for a loop).

nitrocode commented:
/test all

cdobbyn commented:
@nitrocode anything else you need from me here? This isn't failing on code but misconfiguration of the summary bot?

nitrocode commented:
@aknysh @Gowiem do you mind taking one more look before merging?
